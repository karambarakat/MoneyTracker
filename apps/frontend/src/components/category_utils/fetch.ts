import { UseQueryResult, useQuery } from '@tanstack/react-query'
import { useMemo } from 'react'

export const iconQuery = async () => {
  const res = await import('./icons')

  return res.default.split('\n') as string[]
}

/**
 *
 * @param whichOne select specific icon or return all icons
 */
export function useIcon(whichOne: string) {
  const res = useQuery({
    queryFn: iconQuery,
    suspense: false,
    queryKey: ['icons'],
    cacheTime: 1000 * 60 * 60 * 24,
  }) as UseQueryResult<string>
  // return data of type string please

  const validIndex = useMemo(() => {
    const target = Number(whichOne)

    if (isNaN(target) || target > 876 || target < 0)
      return new Error('Invalid index')

    return target
  }, [whichOne])

  if (validIndex instanceof Error) return validIndex

  if (res.status === 'success')
    return { status: 'success', data: res.data[validIndex] }

  return res
}

export const defaultValue =
  'm38.49,32h-14.49V8h24v18h-6.04l-3.46,6Zm-18.49,4v-14c-7.73,0-14,6.27-14,14s6.27,14,14,14,14-6.27,14-14h-14Zm27.73-6h-3.46l-12.12,21c.77,1.33.96,1.67,1.73,3h24.25c.77-1.33.96-1.67,1.73-3l-12.12-21Z'
