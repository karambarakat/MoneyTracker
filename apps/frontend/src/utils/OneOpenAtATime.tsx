import { useId } from '@mantine/hooks'
import { createContext, useContext, useEffect, useState } from 'react'
import { WithChildren } from 'ui/src/utils/WithChildren'

const context = createContext<{
  open: false | string
  setOpen: (s: string | false) => void
}>({ open: false, setOpen: () => console.error('no context assigned') })

export const useOneState = (initial?: boolean) => {
  const id = useId()
  const { open, setOpen } = useContext(context)
  const [_open, _setOpen] = useState(open === id)
  useEffect(() => {
    if (initial) setOpen(id)
  }, [initial])
  useEffect(() => {
    if (open === id) _setOpen(true)
    else _setOpen(false)
  }, [open])

  return [
    _open,
    function (arg: boolean | ((old: boolean) => boolean)) {
      if (typeof arg === 'function') {
        const rt = arg(_open)
        return rt ? setOpen(id) : setOpen(false)
      }
      if (arg) setOpen(id)
      else setOpen(false)
    },
  ] as const
}

export const OneStateProvider = ({ children }: WithChildren) => {
  const [open, setOpen] = useState<false | string>(false)
  return (
    <context.Provider value={{ open, setOpen }}>{children}</context.Provider>
  )
}
