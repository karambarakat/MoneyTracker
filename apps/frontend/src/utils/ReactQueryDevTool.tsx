import { ReactQueryDevtools } from '@tanstack/react-query-devtools'
import React from 'react'

const ReactQueryDevtoolsProductionLazy = React.lazy(() =>
  import('@tanstack/react-query-devtools/build/lib/index.prod.js').then(d => ({
    default: d.ReactQueryDevtools,
  })),
)

export function ReactQueryDevtoolsProduction() {
  const [showDevtools, setShowDevtools] = React.useState(false)

  React.useEffect(() => {
    // @ts-ignore
    window.toggleDevtools = () => setShowDevtools(old => !old)
  }, [])

  return showDevtools ? (
    <React.Suspense fallback={null}>
      <ReactQueryDevtoolsProductionLazy />
    </React.Suspense>
  ) : null
}

export default function ReactQueryDevTools_() {
  return (
    <>
      <ReactQueryDevtools initialIsOpen={false} position="bottom-right" />
      <ReactQueryDevtoolsProduction />
    </>
  )
}
