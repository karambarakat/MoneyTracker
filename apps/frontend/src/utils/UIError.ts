export default class UIError extends Error {
  jsError: unknown
  constructor(message: string, jsError?: unknown) {
    super(message)
    this.jsError = jsError
  }
}
