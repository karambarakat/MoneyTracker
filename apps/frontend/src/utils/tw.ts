// this is temprorary until I switch to UnoCSS or
// figure out a way to simplify tw configuration completely

import tw from 'twin.macro'

export const iconRed = tw`fill-red-800 dark:fill-red-400`

export const iconBlue = tw`fill-blue-800 dark:fill-blue-400`
