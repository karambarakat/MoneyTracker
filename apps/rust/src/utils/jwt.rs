#[allow(dead_code)]
// for some limitation for ts_rs, this exists for the sake of simplicity
#[derive(ts_rs::TS)]
#[ts(export)]
pub struct Jwt_ {
    pub id: String,
    pub email: String,
    pub user_id: String,
    pub exp: String,
    pub iat: String,
}

pub struct Jwt {
    pub id: String,
    pub email: String,
    pub user_id: String,
    pub exp: DateTime<Utc>,
    pub iat: DateTime<Utc>,
}
use chrono::DateTime;
use chrono::Duration;
use chrono::Utc;
use hmac::{Hmac, Mac};
use jwt::SignWithKey;
use jwt::VerifyWithKey;
use sha2::Sha256;
use std::collections::BTreeMap;

use crate::errors::MyErrors;

impl Jwt {
    pub fn validate(str: &str) -> Result<Self, MyErrors> {
        let env = std::env::var("JWT_SALT").expect("JWT_SALT is not set");

        let key: Hmac<Sha256> = Hmac::new_from_slice(env.as_bytes())
            .map_err(|err| MyErrors::InternalError(Box::new(err)))?;

        let claims: BTreeMap<String, String> = str
            .verify_with_key(&key)
            .map_err(|err| MyErrors::InternalError(Box::new(err)))?;

        let exp = claims.get("exp");

        let exp = match exp {
            Some(exp) => {
                // test if expirted
                let exp = exp
                    .parse::<i64>()
                    .map_err(|err| MyErrors::InternalError(Box::new(err)))?;
                let exp = DateTime::<Utc>::from_utc(
                    chrono::NaiveDateTime::from_timestamp_opt(exp, 0).unwrap(),
                    Utc,
                );

                if exp < Utc::now() {
                    return Err(MyErrors::ExpiredBearerToken);
                }

                exp
            }
            None => return Err(MyErrors::ExpiredBearerToken),
        };

        Ok(Self {
            id: claims.get("id").unwrap().to_string(),
            email: claims.get("email").unwrap().to_string(),
            user_id: claims.get("user_id").unwrap().to_string(),
            exp,
            iat: DateTime::<Utc>::from_utc(
                chrono::NaiveDateTime::from_timestamp_opt(
                    claims
                        .get("iat")
                        .unwrap()
                        .parse::<i64>()
                        .map_err(|err| MyErrors::InternalError(Box::new(err)))?,
                    0,
                )
                .unwrap(),
                Utc,
            ),
        })
    }

    pub fn sign(id: &str, email: &str, user_id: &str) -> Result<String, MyErrors> {
        let env = std::env::var("JWT_SALT").expect("JWT_SALT is not set");

        let key: Hmac<Sha256> = Hmac::new_from_slice(env.as_bytes())
            .map_err(|err| MyErrors::InternalError(Box::new(err)))?;

        let claims = {
            let mut claims: BTreeMap<String, String> = BTreeMap::new();

            let now = Utc::now();
            let exp = now + Duration::days(1);

            claims.insert("id".to_string(), id.to_string());
            claims.insert("email".to_string(), email.to_string());
            claims.insert("user_id".to_string(), user_id.to_string());
            claims.insert("exp".to_string(), exp.timestamp().to_string());
            claims.insert("iat".to_string(), now.timestamp().to_string());

            claims
        };

        let token_str = claims
            .sign_with_key(&key)
            .map_err(|err| MyErrors::InternalError(Box::new(err)))?;

        Ok(token_str)
    }
}

impl std::fmt::Display for Jwt {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "[User: {},****]", self.email)
    }
}
