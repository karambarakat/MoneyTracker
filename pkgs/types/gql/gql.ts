/* eslint-disable */
import * as types from './graphql';
import { TypedDocumentNode as DocumentNode } from '@graphql-typed-document-node/core';

/**
 * Map of all GraphQL operations in the project.
 *
 * This map has several performance disadvantages:
 * 1. It is not tree-shakeable, so it will include all operations in the project.
 * 2. It is not minifiable, so the string of a GraphQL query will be multiple times inside the bundle.
 * 3. It does not support dead code elimination, so it will add unused operations.
 *
 * Therefore it is highly recommended to use the babel or swc plugin for production.
 */
const documents = {
    "\n  fragment Category on Category {\n    id\n    title\n    color\n    icon\n  }\n": types.CategoryFragmentDoc,
    "\n  fragment User on User {\n    avatar\n    createdAt\n    displayName\n    email\n    id\n    providers\n    updatedAt\n  }\n": types.UserFragmentDoc,
    "\n  fragment Entry on Entry {\n    id\n    title\n    amount\n    note\n    createdAt\n    updatedAt\n    category {\n      id\n      title\n      color\n      icon\n    }\n  }\n": types.EntryFragmentDoc,
};

/**
 * The graphql function is used to parse GraphQL queries into a document that can be used by GraphQL clients.
 *
 *
 * @example
 * ```ts
 * const query = graphql(`query GetUser($id: ID!) { user(id: $id) { name } }`);
 * ```
 *
 * The query argument is unknown!
 * Please regenerate the types.
 */
export function graphql(source: string): unknown;

/**
 * The graphql function is used to parse GraphQL queries into a document that can be used by GraphQL clients.
 */
export function graphql(source: "\n  fragment Category on Category {\n    id\n    title\n    color\n    icon\n  }\n"): (typeof documents)["\n  fragment Category on Category {\n    id\n    title\n    color\n    icon\n  }\n"];
/**
 * The graphql function is used to parse GraphQL queries into a document that can be used by GraphQL clients.
 */
export function graphql(source: "\n  fragment User on User {\n    avatar\n    createdAt\n    displayName\n    email\n    id\n    providers\n    updatedAt\n  }\n"): (typeof documents)["\n  fragment User on User {\n    avatar\n    createdAt\n    displayName\n    email\n    id\n    providers\n    updatedAt\n  }\n"];
/**
 * The graphql function is used to parse GraphQL queries into a document that can be used by GraphQL clients.
 */
export function graphql(source: "\n  fragment Entry on Entry {\n    id\n    title\n    amount\n    note\n    createdAt\n    updatedAt\n    category {\n      id\n      title\n      color\n      icon\n    }\n  }\n"): (typeof documents)["\n  fragment Entry on Entry {\n    id\n    title\n    amount\n    note\n    createdAt\n    updatedAt\n    category {\n      id\n      title\n      color\n      icon\n    }\n  }\n"];

export function graphql(source: string) {
  return (documents as any)[source] ?? {};
}

export type DocumentType<TDocumentNode extends DocumentNode<any, any>> = TDocumentNode extends DocumentNode<  infer TType,  any>  ? TType  : never;