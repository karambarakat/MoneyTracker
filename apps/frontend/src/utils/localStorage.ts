import { useEffect, useState } from 'react'
import { Jwt_ } from 'types/backend'
import { User } from 'types/gql/graphql'

const eventTarget = new EventTarget()

class LocalItem<T> {
  /**
   * @private
   */
  _item: T | undefined
  key: string
  constructor(key: string) {
    this.key = key
    const localItem = localStorage.getItem(key)
    this._item = localItem ? JSON.parse(localItem as string) : undefined
  }

  setItem(item: T | undefined) {
    if (!item) localStorage.removeItem(this.key)
    else localStorage.setItem(this.key, JSON.stringify(item))

    this._item = item

    // React
    eventTarget.dispatchEvent(new Event(this.key))
  }

  getItem() {
    const item = localStorage.getItem(this.key)

    if (!item) return

    return JSON.parse(item) as T
  }

  use() {
    const [item, setItem] = useState<LocalItem<T>>(this)

    useEffect(() => {
      const handler = () => setItem(new LocalItem(this.key))
      eventTarget.addEventListener(this.key, handler)
      return () => eventTarget.removeEventListener(this.key, handler)
    }, [])

    return item
  }
}

export const profile = new LocalItem<User>('profile')

export const token = Object.assign(new LocalItem<string>('token'), {
  valid(this: LocalItem<string>) {
    const item = this.getItem()

    if (!item) return false

    return (
      Number((JSON.parse(atob(item.split('.')[1])) as Jwt_).exp) >=
      Date.now() / 1000
    )
  },
})
