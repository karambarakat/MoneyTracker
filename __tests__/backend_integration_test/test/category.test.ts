import { afterAll, beforeAll, expect, it } from '@jest/globals'
import { fetchQql, gql } from '../utils/gql'
import fetch from 'node-fetch'
import { HELPER_SERVER, REST_API } from '../utils/constants'
import { categoryGQL } from '../utils/expect'
import {
  Mutation,
  MutationCreateOneCategoryArgs,
  MutationCreateManyCategoriesArgs,
  MutationDeleteOneCategoryArgs,
  MutationDeleteManyCategoriesArgs,
  MutationUpdateOneCategoryArgs,
} from 'types/gql/graphql'

const ctx: any = {
  ids: [],
  credential: 'testing_category@m.c:pass',
  credential2: 'testing_category2@m.c:pass2',
}

beforeAll(async () => {
  const res = await fetch(REST_API + '/auth/local/register', {
    method: 'POST',
    headers: {
      content: 'application/json',
      authorization: 'Basic ' + Buffer.from(ctx.credential).toString('base64'),
    },
    body: JSON.stringify({}),
  })
  const json: any = await res.json()
  ctx.user = Object.assign({}, json.data)
  ctx.ids.push(json.data.id)
  ctx.user.token = res.headers.get('x-token')
  ctx.user.headers = {
    authorization: 'Bearer ' + ctx.user.token,
  }

  const res2 = await fetch(REST_API + '/auth/local/register', {
    method: 'POST',
    headers: {
      content: 'application/json',
      authorization: 'Basic ' + Buffer.from(ctx.credential2).toString('base64'),
    },
    body: JSON.stringify({}),
  })

  const json2: any = await res2.json()
  ctx.user2 = Object.assign({}, json2)
  ctx.ids.push(json2.id)
  ctx.user2.token = res2.headers.get('x-token')
  ctx.user2.headers = {
    authorization: 'Bearer ' + ctx.user2.token,
  }
})

afterAll(async () => {
  await Promise.all(
    ctx.ids.map((id: string) =>
      fetch(HELPER_SERVER + '/remove_user_data/' + id),
    ),
  )
})

it('create_one_category', async () => {
  const input = gql`
    mutation create_one_category($category: CategoryInput!) {
      createOneCategory(category: $category) { ${gql.category} }
    }
  `

  const vars = {
    category: {
      title: 'new_category',
    },
  } satisfies MutationCreateOneCategoryArgs

  const data = await fetchQql<
    {
      createOneCategory: Mutation['createOneCategory']
    },
    MutationCreateOneCategoryArgs
  >(ctx.user.headers, input.input, vars)

  ctx.cat1 = data.createOneCategory
  ctx.cat1.vars = vars.category

  expect(data.createOneCategory).toMatchSnapshot(categoryGQL)
})

it('create_many_categories', async () => {
  const input = gql`
    mutation create_many_categories($categories: [CategoryInput!]!) {
      createManyCategories(categories: $categories) { ${gql.category} }
    }
  `

  const vars = {
    categories: [
      {
        title: 'new_category2',
        color: 'red',
        icon: 'icon',
      },
      {
        title: 'new_category3',
        color: 'blue',
        icon: 'icon_2',
      },
    ],
  } satisfies MutationCreateManyCategoriesArgs

  const data = await fetchQql<
    {
      createManyCategories: Mutation['createManyCategories']
    },
    MutationCreateManyCategoriesArgs
  >(ctx.user.headers, input.input, vars)

  expect(data.createManyCategories.length).toBe(2)

  ctx.cat2 = data.createManyCategories[0]
  ctx.cat3 = data.createManyCategories[1]
  ctx.cat2.vars = vars.categories[0]
  ctx.cat3.vars = vars.categories[1]

  expect(data.createManyCategories[0]).toMatchSnapshot(categoryGQL)

  const data2 = await fetchQql(ctx.user.headers, '{ getAllCategories { id } }')

  expect(data2.getAllCategories.length).toBe(3)
  expect(data2.getAllCategories[0].id).toBe(ctx.cat1.id)
  expect(data2.getAllCategories[1].id).toBe(ctx.cat2.id)
  expect(data2.getAllCategories[2].id).toBe(ctx.cat3.id)
})

it('create results', async () => {
  const data = await fetchQql(
    ctx.user.headers,
    '{ getAllCategories { id title color icon createdBy { id } } }',
  )
  expect(data.getAllCategories.length).toBe(3)
  expect(data.getAllCategories[1].id).toBe(ctx.cat2.id)
  expect(data.getAllCategories[1].title).toBe(ctx.cat2.vars.title)
  expect(data.getAllCategories[1].color).toBe(ctx.cat2.vars.color)
  expect(data.getAllCategories[1].icon).toBe(ctx.cat2.vars.icon)
  expect(data.getAllCategories[1].createdBy.id).toBe(ctx.user.id)

  const data2 = await fetchQql(
    ctx.user.headers,
    `{ getOneCategory(id: "${ctx.cat1.id}") { ${gql.category} } }`,
  )

  expect(data2.getOneCategory).toMatchSnapshot(categoryGQL)

  const data3 = await fetchQql(ctx.user2.headers, '{ getAllCategories { id } }')

  expect(data3.getAllCategories.length).toBe(0)
})

it('update_one_category', async () => {
  const input = gql`
    mutation update_one_category($id: ID!, $category: CategoryInput!) {
      updateOneCategory(id: $id, category: $category)
    }
  `
  const vars = {
    id: ctx.cat1.id,
    category: {
      title: 'new_title',
      color: 'new_color',
      icon: 'new_icon',
    },
  } satisfies MutationUpdateOneCategoryArgs

  const data = await fetchQql<
    {
      updateOneCategory: Mutation['updateOneCategory']
    },
    MutationUpdateOneCategoryArgs
  >(ctx.user.headers, input.input, vars)

  expect(data.updateOneCategory).toBe(true)

  const data2 = await fetchQql(
    ctx.user.headers,
    `{ getOneCategory(id: "${ctx.cat1.id}") { ${gql.category} } }`,
  )

  expect(data2.getOneCategory).toMatchSnapshot(categoryGQL)
  expect(data2.getOneCategory.title).toBe('new_title')
  expect(data2.getOneCategory.color).toBe('new_color')
  expect(data2.getOneCategory.icon).toBe('new_icon')
})

it('delete_one_category', async () => {
  const input = gql`
    mutation delete_one_category($id: ID!) {
      deleteOneCategory(id: $id)
    }
  `

  const vars = {
    id: ctx.cat1.id,
  } satisfies MutationDeleteOneCategoryArgs

  const data = await fetchQql<
    {
      deleteOneCategory: Mutation['deleteOneCategory']
    },
    MutationDeleteOneCategoryArgs
  >(ctx.user.headers, input.input, vars)

  expect(data.deleteOneCategory).toBe(true)

  const data2 = await fetchQql(ctx.user.headers, '{ getAllCategories { id } }')

  expect(data2.getAllCategories.length).toBe(2)
  expect(data2.getAllCategories[0].id).toBe(ctx.cat2.id)
  expect(data2.getAllCategories[1].id).toBe(ctx.cat3.id)
})

it('delete_many_categories', async () => {
  const input = gql`
    mutation delete_many_categories($ids: [ID!]!) {
      deleteManyCategories(ids: $ids)
    }
  `

  const vars = {
    ids: [ctx.cat2.id, ctx.cat3.id],
  } satisfies MutationDeleteManyCategoriesArgs

  const data = await fetchQql<
    {
      deleteManyCategories: Mutation['deleteManyCategories']
    },
    MutationDeleteManyCategoriesArgs
  >(ctx.user.headers, input.input, vars)

  expect(data.deleteManyCategories).toBe(2)

  const data2 = await fetchQql(ctx.user.headers, '{ getAllCategories { id } }')

  expect(data2.getAllCategories.length).toBe(0)
})
