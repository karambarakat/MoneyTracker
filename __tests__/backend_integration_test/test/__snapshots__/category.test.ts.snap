// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create results 1`] = `
Object {
  "color": null,
  "createdAt": Any<String>,
  "createdBy": Object {
    "avatar": null,
    "createdAt": Any<String>,
    "displayName": null,
    "email": "testing_category@m.c",
    "id": Any<String>,
    "providers": "local",
    "updatedAt": Any<String>,
  },
  "icon": null,
  "id": Any<String>,
  "title": "new_category",
  "updatedAt": Any<String>,
}
`;

exports[`create_many_categories 1`] = `
Object {
  "color": "red",
  "createdAt": Any<String>,
  "createdBy": Object {
    "avatar": null,
    "createdAt": Any<String>,
    "displayName": null,
    "email": "testing_category@m.c",
    "id": Any<String>,
    "providers": "local",
    "updatedAt": Any<String>,
  },
  "icon": "icon",
  "id": Any<String>,
  "title": "new_category2",
  "updatedAt": Any<String>,
  "vars": Object {
    "color": "red",
    "icon": "icon",
    "title": "new_category2",
  },
}
`;

exports[`create_one_category 1`] = `
Object {
  "color": null,
  "createdAt": Any<String>,
  "createdBy": Object {
    "avatar": null,
    "createdAt": Any<String>,
    "displayName": null,
    "email": "testing_category@m.c",
    "id": Any<String>,
    "providers": "local",
    "updatedAt": Any<String>,
  },
  "icon": null,
  "id": Any<String>,
  "title": "new_category",
  "updatedAt": Any<String>,
  "vars": Object {
    "title": "new_category",
  },
}
`;

exports[`update_one_category 1`] = `
Object {
  "color": "new_color",
  "createdAt": Any<String>,
  "createdBy": Object {
    "avatar": null,
    "createdAt": Any<String>,
    "displayName": null,
    "email": "testing_category@m.c",
    "id": Any<String>,
    "providers": "local",
    "updatedAt": Any<String>,
  },
  "icon": "new_icon",
  "id": Any<String>,
  "title": "new_title",
  "updatedAt": Any<String>,
}
`;
